modules/git/repo_base.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+import (
+	"context"
+	"io"
+)
+
+// contextKey is a value for use with context.WithValue.
+type contextKey struct {
+	name string
+}
+
+// RepositoryContextKey is a context key. It is used with context.Value() to get the current Repository for the context
+var RepositoryContextKey = &contextKey{"repository"}
+
+// RepositoryFromContext attempts to get the repository from the context
+func RepositoryFromContext(ctx context.Context, path string) *Repository {
+	value := ctx.Value(RepositoryContextKey)
+	if value == nil {
+		return nil
+	}
+
+	if repo, ok := value.(*Repository); ok && repo != nil {
+		if repo.Path == path {
+			return repo
+		}
+	}
+
+	return nil
+}
+
+type nopCloser func()
```

Comment:
I'm aware of type X struct{} - does a func() type have any specific means?

Reply:
it's essentially the same as struct{} here except that it should more easily get inlined to a noop.

However, say instead you wanted to wrap some function to become an io.Closer, you could write:

```go
type asCloser func() error

func (a asCloser) Close() error {
  return a()
} 
```

and avoid the wrapper struct.